README
NAME
    Mojolicious::Plugin::CaptchaRenderer - captcha renderer for Mojolicious
    framework

VERSION
    0.01

SYNOPSIS
       # Mojolicious::Lite
       plugin captcha_renderer => { size => 20, color => 'blue', wave_amplitude => 4};
       get '/img/code.png' => sub {
          my $self = shift;
          $self->render(handler => 'captcha', code => 'cool captcha code');
       }
   
       # Mojolicious
       $self->plugin(captcha_renderer => { size => 20, color => 'blue', wave_amplitude => 4});
   
       package MyApp::MyController;
   
       sub my_action {
          my $self = shift;
          $self->render(handler => 'captcha', code => 'cool captcha code');
       }

OPTIONS
    size - font size. By default 60
    color - font color. By default black
    font - font name. By default undef
    bgcolor - captcha background color. By default white
    wave_amplitude - amplitude of wave in captcha. By default 7
    wave_length - length of wave in captcha. By dafault 80

SUPPORT
    *   Repository

        <http://github.com/konstantinov/Mojolicious-Plugin-CaptchaRenderer>

SEE ALSO
    Mojolicious, Mojolicious::Plugin, Mojolicious::Lite

COPYRIGHT & LICENSE
    Copyright 2010 Dmitry Konstantinov. All right reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl module Mojolicious::Plugin::CaptchaRenderer

Mojolicious::Plugin::CaptchaRenderer - captcha renderer for Mojolicious framework

0.02 2011-02-10

  * updating to Mojolicious 1.0 capatibility

0.01 2010-07-13

  * First release to CPAN.


demo/captcha.pl
#!/usr/bin/perl

use strict;
use lib qw(../lib);

use Mojolicious::Lite;

my $CAPTCHA;

plugin captcha_renderer => { size => 50 };

any '/' => sub {
	my $self = shift;
	
	if ($self->req->param('submit')) {
		$self->check::form;
		
		return $self->render(ok => 1) 
			unless %{$self->stash('error')}
		;
	}
} => 'index';

get '/images/code.png' => sub {
	my $self = shift;
	
	my @letter = (0..9, 'a'..'z');
	
	my $code = join '', @letter[ map { int rand @letter } 1..5 ]; 
	$CAPTCHA->{lc $code} = 1;
	$self->render_data($self->captcha($code));
};

app->start;

sub check::form {
	my $self = shift;
	
	$self->stash(error => {
		(map { $_ => 'Empty'       } grep { !$self->req->param($_) } qw(name address captcha)),
		(map { $_ => 'Bad captcha' } grep {  $self->req->param($_) && ! delete $CAPTCHA->{lc $self->req->param($_)} } qw(captcha)),
	});
}

__DATA__
@@ index.html.ep
% my $e = $self->stash('error');
<html>
	<head>
		<title>Mojolicious::Plugin::CaptchaRenderer Demo</title>
	</head>
	<body>
		<% unless ($self->stash('ok')) { %>
		<h1>Test form</h1>
		<form action="/" method="post">
			<input name="name" value="<%= param('name') || 'Enter your name' %>" /><br />
			<%== "<b>$e->{name}</b><br/>" if $e->{name}; %>
			
			<input name="address" value="<%= param('address') || 'Enter your address' %>" /><br />
			<%== "<b>$e->{address}</b><br/>" if $e->{address}; %>
			
			<img src="/images/code.png" alt="captcha" /><input name="captcha" value="&lt;-- enter here" /><br />
			<%== "<b>$e->{captcha}</b><br/>" if $e->{captcha}; %>
			<input type="submit" value="Send form" name="submit" />
		</form>
		<% } else { %>
		<h1>Congratulations! Captcha test passed!</h1>
		<% } %>
	</body>
</html>

lib/Mojolicious/Plugin/Captcha.pm
package Mojolicious::Plugin::Captcha;

use strict;
use warnings;
use File::Temp;

use base 'Mojolicious::Plugin';

BEGIN {
	die 'Module Image::Magick not properly installed' unless eval { require Image::Magick; 1 }
}

sub register {
	my ($self,$app,$conf) = @_;
	
	$conf = {
		color          => 'black',
		bgcolor        => 'white',
		size           => 60,
		fond           => 'sans-serif',
		wave_amplitude => 7,
		wave_length    => 80,
		%$conf,
	};
	
	$app->renderer->add_handler(
		captcha => sub {
			my ($r,$c,$output,$options) = @_;
			
			$$output = captcha($c,$conf);
		}
	);
}

sub captcha {
	my ($c,$conf) = @_;
	
	my $img = Image::Magick->new(size => '400x400', magick => 'png');
	my $x; $x = $img->Read('gradient:#ffffff-#ffffff');
	
	$x = $img->Annotate(
		pointsize   => $conf->{'size'}, 
		fill        => $conf->{'color'}, 
		text        => $c->stash('code'), 
		geometry    => '+0+' . $conf->{'size'},
		font        => $conf->{'font'},
	);
	
	warn $x if $x;
	$x = $img->Wave(amplitude => $conf->{'wave_amplitude'}, wavelength => $conf->{'wave_length'});
	warn $x if $x;
	$x = $img->Trim;
	warn $x if $x;
	
	my $body = '';
	
	{
		my $fh = File::Temp->new(UNLINK => 1, DIR => $ENV{'MOJO_TMPDIR'});
		$x = $img->Write('png:' . $fh->filename);
		open $fh, '<', $fh->filename;
		local $/;
		$body = <$fh>;
	}
	warn $x if $x;
	
	
	$c->tx->res->headers->content_type('image/png');
	return $body;
}

1;

lib/Mojolicious/Plugin/CaptchaRenderer.pm
package Mojolicious::Plugin::CaptchaRenderer;

use strict;
use warnings;
use File::Temp;
use File::Spec;

use base 'Mojolicious::Plugin';

BEGIN {
	die 'Module Image::Magick not properly installed' unless eval { require Image::Magick; 1 }
}

our $VERSION = 0.01;

sub register {
	my ($self,$app,$conf) = @_;
	
	$conf = {
		color          => 'black',
		bgcolor        => 'white',
		size           => 60,
		wave_amplitude => 7,
		wave_length    => 80,
		%$conf,
	};
	
	$app->renderer->add_handler(
		captcha => sub {
			my ($r,$c,$output,$options) = @_;
			
			$$output = captcha($c,$conf);
		}
	);
}

sub captcha {
	my ($c,$conf) = @_;
	
	my $img = Image::Magick->new(size => '400x400', magick => 'png');
	my $x; $x = $img->Read('gradient:#ffffff-#ffffff');
	
	$x = $img->Annotate(
		pointsize   => $conf->{'size'}, 
		fill        => $conf->{'color'}, 
		text        => $c->stash('code'), 
		geometry    => '+0+' . $conf->{'size'},
		$conf->{'font'} ? (font => $conf->{'font'}) : (),
	);
	
	warn $x if $x;
	$x = $img->Wave(amplitude => $conf->{'wave_amplitude'}, wavelength => $conf->{'wave_length'});
	warn $x if $x;
	$x = $img->Trim;
	warn $x if $x;
	
	my $body = '';
	
	{
		my $fh = File::Temp->new(UNLINK => 1, DIR => $ENV{MOJO_TMPDIR} || File::Spec->tmpdir);
		$x = $img->Write('png:' . $fh->filename);
		open $fh, '<', $fh->filename;
		local $/;
		$body = <$fh>;
	}
	warn $x if $x;
	
	
	$c->tx->res->headers->content_type('image/png');
	return $body;
}

1;

=head1 NAME

Mojolicious::Plugin::CaptchaRenderer - captcha renderer for Mojolicious framework

=head1 VERSION

0.01

=head1 SYNOPSIS

   # Mojolicious::Lite
   plugin captcha_renderer => { size => 20, color => 'blue', wave_amplitude => 4};
   get '/img/code.png' => sub {
      my $self = shift;
      $self->render(handler => 'captcha', code => 'cool captcha code');
   }
   
   # Mojolicious
   $self->plugin(captcha_renderer => { size => 20, color => 'blue', wave_amplitude => 4});
   
   package MyApp::MyController;
   
   sub my_action {
      my $self = shift;
      $self->render(handler => 'captcha', code => 'cool captcha code');
   }
   
=head1 OPTIONS

=over 4

=item size - font size. By default 60

=item color - font color. By default black

=item font - font name. By default undef

=item bgcolor - captcha background color. By default white

=item wave_amplitude - amplitude of wave in captcha. By default 7

=item wave_length - length of wave in captcha. By dafault 80

=back

=head1 SUPPORT

=over 4

=item * Repository

L<http://github.com/konstantinov/Mojolicious-Plugin-CaptchaRenderer>

=back

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Plugin>, L<Mojolicious::Lite>

=head1 COPYRIGHT & LICENSE

Copyright 2010 Dmitry Konstantinov. All right reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.